Homepage news items section rework

luistoptal opened the pull request:
# Pull request for issue: #208

NOTE: I suggest to review this PR and https://github.com/gigascience/gigadb-website/pull/2167 simultaneously since #2167 includes refactoring of the carousel added in this PR, mainly the carousel is abstracted to a reusable partial

This is a pull request for the following functionalities:

* full rework of news section in home page

![image](https://github.com/user-attachments/assets/2c99becb-a95d-41c6-83cc-43aa59be7b90)

## How to test?

Visit the home page and verify the following:

- I expect there are no news items to display, so the section should not be displayed at all
- go to protected/components/NewsAndFeedsService.php and in the getTodaysNews function remove the second conditional so that all news are displayed
- refresh the home page. you should see a news section with a carousel and news boxes
- clicking on a "read more" link should open the detail page

alternatively you can create more news items from the admin page http://gigadb.gigasciencejournal.com/news/admin and they should show up in the home page

## How have functionalities been implemented?

- the carousel uses the bootstrap carousel and overrides the styles to match the gigadb colors and to make it look better
- the logic was tweaked to cover a handful of use cases:
  - the carousel is responsive and reactive to the screen width
  - mobile shows one item
  - medium screen shows two items
  - desktop shows three items
  - if the number of items to display is less than the maximum, for example, for desktop, if there are just two news items, rather than showing a blank space, the two news items fill all the space
  - if there are less news items than the maximum that can be displayed at any given time (e.g. 3 for desktop) then the carousel controls are hidden

## Any issues with implementation?

--

## Any changes to automated tests?

tests pass. As far as I have seen, there are no acceptance tests covering this case

luistoptal committed:
feat: replaces news section by responsive carousel of boxes

luistoptal committed:
refactor: adds styles to separate less file

luistoptal committed:
refactor: nests less classes

luistoptal committed:
feat: clean up news block design

luistoptal committed:
feat: enhances web accessibility

luistoptal committed:
feat: dynamically adjust visibility of carousel controls

luistoptal committed:
feat: dynamically update max width depending on number of news items to display

luistoptal committed:
feat: increase excerpt length

luistoptal committed:
refactor: move helper outside jquery wrapper since it uses pure js

luistoptal committed:
chore: up log

luistoptal committed:
chore: Merge branch 'develop' into new-feature/208-news-items-on-the-home-page

only1chunts commented:
@luistoptal what is the situation with this work? is the PR ready for review? I ask because I've just added items to the news feed on live and realised that the work hasn't been released yet.

luistoptal commented:
Hi @only1chunts most of my PRs (if not all) are ready for review, I keep them as drafts to avoid having lots of PRs for review at one time, then I set new for review as the existing ones are merged

I try to keep 3-5 maximum open for review at any given time

rija commented:
Hi @luistoptal,

We have a Github project  to manage PRs now.
if a PR of yours is ready for scrutiny and is rebased with develop's HEAD, feel free to open it for review and move it to the "Ready for review".

It's not going to overwhelm the core team as we pick next PRs to review from the "Reviewing required" column which is deliberately capped in capacity.

This should allow draft PRs to be just that, drafts -as in unfinished work-.

Moving PRs from "Ready for review"  to "Reviewing required" is something the core team do based on how well we are progressing with the backlog and on liaising with product owner for when an issue with a ready PR has change in priority.

Also, from now on we probably should not close issues who have a PR attached still in draft mode.

luistoptal committed:
chore: Merge branch 'develop' into new-feature/208-news-items-on-the-home-page

luistoptal commented:
Hi @rija , perfect, then I will move all my ready PRs to ready for review

pli888 reviewed (changes requested):
**praise**: Code is working on my dev and staging environments.
**chore**: Need to rebase with develop branch to resolve conflict with CHANGELOG.md
**issue**: Sometimes, curators add hyperlinks into news item text in the form of <a> tags but these are not displayed as links on home page in news carousel.
**thought**: I think its best to include an acceptance test for creating and displaying news. Maybe something like this is a start:

```
Feature: News items
  As a curator
  I want to access a web form to create news items
  So that I can display news on gigadb.org home page

  Background:
    Given I have signed in as admin

  @ok
  Scenario: Logging in into gigadb website with the credentials of an existing account
    Given I am on "/news/create"
    When I fill in the field of "name" "News[title]" with "Hello World"
    And I fill in the field of "name" "News[body]" with "Some stuff goes here."
    And I fill in the field of "name" "News[start_date]" with "2025-05-02"
    And I fill in the field of "name" "News[end_date]" with "2035-05-02"
    And I press the button "Create"
    And I am on "/"
    Then I should see "Some stuff goes here."
```

luistoptal committed:
feat: purify news content

luistoptal committed:
feat: simplify news body line clamp using css

luistoptal committed:
chore: test news creation

luistoptal commented:
Hi @pli888 

I have added a new test as per your suggestions `tests/acceptance/AdminNews.feature`, I created a new test file based on similar other test files

As for this:

> Sometimes, curators add hyperlinks into news item text in the form of tags but these are not displayed as links on home page in news carousel

I have updated the html purification to allow <a> elements. Other "dangerous" elements are parsed out. This could be tested by creating a news with this body:

```
A safe link: <a href="https://example.com">Example</a>  
An inline script: <script>alert('XSS');</script>  
An image with JS: <img src="x" onerror="alert('XSS')">  
Another link: <a href="javascript:alert('XSS')">Click me</a>  
Bold text: <b>Bold</b>  
Link with target: <a href="https://safe.com" target="_blank" rel="noopener">Safe</a>  
Div tag: <div onclick="alert('XSS')">Click me</div>
```

links should be visible, but no "alert" should pop up, and the display will look like this:

![image](https://github.com/user-attachments/assets/8119fee0-312d-4b9e-893f-c7f7c1fcab04)


I have also added this to the news detail page

![image](https://github.com/user-attachments/assets/201a22e3-8499-4975-b8ef-4dabc8e6b6ff)


As a consequence of parsing html, the programmatic text truncation became too complicated since I would have to parse the text such that the truncation still generated valid html (doable but very overkill), so I switched to a much simpler CSS-only text truncation that clamps the number of lines to 4, i.e.:

```css
      .news-body {
        flex-grow: 1;
        margin-bottom: 16px;

        color: @darker-gray-on-white;
        font-size: 0.95rem;
        line-height: 1.5;

        // multiline line-clamp
        display: -webkit-box;
        -webkit-box-orient: vertical;
        -webkit-line-clamp: 4;
        line-clamp: 4;
        overflow: hidden;
        text-overflow: ellipsis;
        max-height: calc(1.5em * 4);
      }
```

So, now, if the curators add html to the news when creating one, safe html (links, bold, italic, etc) will be parsed correctly

Examples:

![image](https://github.com/user-attachments/assets/54951dc7-32b8-43a9-8e35-9794efcb2113)

![image](https://github.com/user-attachments/assets/34a81935-53b9-4505-9533-b758014ab358)

However, news that are already present show escaped characters such as `&lt;` which are used to show things like `<` without having them parsed as html. In such cases there is no option but to leave them as text as these shouldn't be parsed. If the same issue arises in production, the only reasonable solution would be to either ask the curators to manually update the text, or to run a one time script to replace the & characters

pli888 reviewed (approved):
**praise**: New functionality to display safe links in news items is working on dev and staging. 
**praise**: Good to have acceptance test for news items now.

Happy to approve.

kencho51 reviewed (approved):
Hi @luis:

**praise:** I can see the news carousel working properly on the homepage after creating a new news from the admin page.
**praise:** The news carousel can parse the html and allows url redirect.
**issue (non-blocking):** For the acceptance tests, the tag `ok-can-offline` was created at first when you have no `GITLAB_PRIVATE_TOKEN` yet, so you can run these acceptance tests locally. But now, you have one and are able to execute the gitlab pipeline, the tags `ok-can-offline` and `ok-needs-secrets` should be removed.
**suggestion (non-blocking):** Remove the tags `ok-can-offline` and `ok-needs-secrets`.
**chore:** Fix the changelog conflicts.
**praise:** All tests are passing.

No blocking issues from me, happy to approve.

luistoptal committed:
chore: remove deprecated test tag

luistoptal committed:
chore: Merge branch 'develop' into new-feature/208-news-items-on-the-home-page

luistoptal commented:
Hi @kencho51 

I fixed marge conflicts and removed tag as requested

kencho51 commented:
Hi @luistoptal,

> 
> I fixed marge conflicts and removed tag as requested

Then you need to refactor  the `gigadb-website/tests/acceptance_runner` as well.

For example: 

```
#!/bin/bash

set -e
set -u

source .env

profile=${1:-"local"}

if [[ $(uname -m) == 'arm64' ]]; then
  # Disable legacy @javascript tagged Behat tests on Apple silicon based macs until (TODO) we migrate those tests to Codeception
  profile="apple-silicon"
fi

 if [ ${GITLAB_PRIVATE_TOKEN:+1} ] ;then
   docker-compose run --rm codecept run --no-redirect acceptance
   docker-compose run --rm application ./protected/yiic migrate --interactive=0
   docker-compose run --rm test ./vendor/behat/behat/bin/behat --profile $profile -v --stop-on-failure
 fi

```

After that, all the acceptance tests should be executed in both dev and CI.

luistoptal commented:
Hi @kencho51 , I only removed the tags from the test file I created, from your last comment I suspect what you meant is to remove all `ok-needs-secrets` and `ok-can-offline` from all test files, is that right? no one else needs the ok-can-offline tag?

kencho51 commented:
Hi @luistoptal,

Since all of us should have `GITLAB_PRIVATE_TOKEN`, I think these tags `ok-needs-secrets` and `ok-can-offline` can be removed, and the test runner should be updated too.

luistoptal committed:
chore: delete all ok-needs-secrets and ok-can-offline tags

luistoptal commented:
Hi @kencho51 I deleted all occurences of the ok-needs-secrets and ok-can-offline tags as you suggest, and updated the acceptance_runner

luistoptal committed:
chore: Merge branch 'develop' into new-feature/208-news-items-on-the-home-page

luistoptal commented:
closing because another PR already incorporates these changes
